### PR TITLE
Initial implementation of GPUBindGroupLayout for WebGPU

### PR DESCRIPTION
**Pull request by imiklos**
Added WebIDL bindings for `GPUBindGroupLayout`, `GPUBindGroupLayoutDescriptor`, `GPUBindingType`, `GPUShaderStage` and `GPUBindGroupLayoutBinding` (Note: The servo's codegen doesn't like the name, because its already occupied).
Implemented the `createBindGroupLayout` function of `GPUDevice`.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes addresses a part of #24706

cc @kvark @jdm @zakorgy

### CODE REVIEW

**Comment by imiklos** [comment]
Added `peek-poke` and `peek-poke-derive` to the ignore list of duplicate versions.

File: components/script/dom/gpubindgrouplayout.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+use dom_struct::dom_struct;
+use std::cell::Cell;
+use webgpu::{WebGPU, WebGPUBindGroupLayout};
+
+#[dom_struct]
+pub struct GPUBindGroupLayout {
+    reflector_: Reflector,
+    label: DomRefCell<Option<DOMString>>,
+    bind_group_layout: WebGPUBindGroupLayout,
+    #[ignore_malloc_size_of = "mozjs"]
```
Comment (jdm) [inline components/script/dom/gpubindgrouplayout.rs]: Use `"defined in webgpu"` here instead.

File: components/script/dom/gpubindgrouplayout.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    bindings: Vec<GPUBindGroupLayoutBindings>,
+    #[ignore_malloc_size_of = "channels are hard"]
```
Comment (jdm) [inline components/script/dom/gpubindgrouplayout.rs]: Same here - this isn't a channel, as far as I know.

File: components/script/dom/gpudevice.rs [hunk 3]
```diff
@@ -229,4 +236,181 @@ impl GPUDeviceMethods for GPUDevice {
...
+        let bindings = descriptor
+            .bindings
+            .iter()
+            .map(|bind| {
+                // TODO: binding must be >= 0
+                storeBindings.insert(bind.binding);
+                let visibility = match ShaderStage::from_bits(bind.visibility) {
+                    Some(visibility) => visibility,
+                    None => {
+                        valid &= false;
```
Comment (jdm) [inline components/script/dom/gpudevice.rs]: This is the same as `valid = false`, right? If so, let's use that instead. Same comment about later uses in this method, too.

File: components/script/dom/gpudevice.rs [hunk 4]
```diff
@@ -229,4 +236,181 @@ impl GPUDeviceMethods for GPUDevice {
...
+                    multisampled: bind.multisampled,
+                    texture_dimension: webgpu::wgpu::resource::TextureViewDimension::D2, // Use as default for now
+                }
+            })
+            .collect::<Vec<BindGroupLayoutBinding>>();
+
+        // bindings are unique
+        valid &= storeBindings.len() == bindings.len();
+
+        // Values does not exceeded the max limit for each ShaderStage
```
Comment (jdm) [inline components/script/dom/gpudevice.rs]: `// Ensure that values do not exceed the max limit for each ShaderStage.`

File: components/script/dom/gpudevice.rs [hunk 5]
```diff
@@ -229,4 +236,181 @@ impl GPUDeviceMethods for GPUDevice {
...
+        for each in validation_map.values() {
```
Comment (jdm) [inline components/script/dom/gpudevice.rs]:
```rust
valid &= validation_map.values().all(|stage| {
    stage.max_uniform_buffers_per_shader_stage >= 0 &&
    ...
});
```

File: components/script/dom/gpudevice.rs [hunk 6]
```diff
@@ -229,4 +236,181 @@ impl GPUDeviceMethods for GPUDevice {
...
+                each.max_sampled_textures_per_shader_stage < 0 ||
+                each.max_storage_textures_per_shader_stage < 0 ||
+                each.max_samplers_per_shader_stage < 0
+            {
+                valid &= false;
+                break;
+            }
+        }
+
+        // DynamicValues does not exceeded the max limit for the pipeline
```
Comment (jdm) [inline components/script/dom/gpudevice.rs]: `does not exceed`

File: components/script/dom/webidls/GPUBindGroupLayout.webidl [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+[Exposed=(Window, DedicatedWorker), Serializable, Pref="dom.webgpu.enabled"]
+interface GPUBindGroupLayout {
+};
+GPUBindGroupLayout includes GPUObjectBase;
+
+dictionary GPUBindGroupLayoutDescriptor : GPUObjectDescriptorBase {
+    required sequence<GPUBindGroupLayoutBindings> bindings;
+};
+
+// Note: Servo codegen doesn't like the name `GPUBindGroupLayoutBinding` because it's already occupied
```
Comment (jdm) [inline components/script/dom/webidls/GPUBindGroupLayout.webidl]: What's the full error message? What's the source of the duplicate name?

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good!

File: components/script/dom/webidls/GPUBindGroupLayout.webidl (hunk 7 above)
Comment (imiklos) [reply components/script/dom/webidls/GPUBindGroupLayout.webidl]: When the codegen generates the internal `Binding` from the WebIDL for the `GPUBindGroupLayout` interface, the internal name for this bind will be `GPUBindGroupLayoutBinding` and when the codegen tries to generate the `GPUBindGroupLayoutBinding` dictionary we got a duplicate name.
The error message: `GPUBindGroupLayoutBinding` must be defined only once in the type namespace of this module.

File: components/script/dom/gpudevice.rs (hunk 3 above)
Comment (imiklos) [reply components/script/dom/gpudevice.rs]: You are right! Thank you!

**Comment by imiklos** [comment]
@jdm Thank you! I've addressed the changes you mentioned and updated the pull request.

**Comment by jdm** [comment]
I have verified that there is no way for the WebIDL dictionary name to be exposed to content, so renaming it seems like a fine solution.
@bors-servo r+

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Pull request merged**